azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/prompt/DefaultPrompter.java

```diff
@@ -0,0 +1,287 @@
+        for (final T entity : entities) {
+            final String displayLine = String.format("%2d. %s", index++, getNameFunc.apply(entity));
+            System.out.println(defaultEntity == entity ? TextUtils.blue(displayLine) : displayLine);
+        }
+
+        final int selectedIndex = entities.indexOf(defaultEntity);
+
+        final String defaultValueMessage = selectedIndex >= 0 ? " (" + TextUtils.blue(new Integer(selectedIndex + 1).toString()) + ")" : "";
+        final String hintMessage = String.format("[1-%d]%s", entities.size(), defaultValueMessage);
+        System.out.printf(PROMPT_MESSAGE_INDEX_TEMPLATE, name, hintMessage);
+        System.out.flush();
+
+        return loopInput(defaultEntity, hasDefaultValue, isRequired, String.format(PROMPT_MESSAGE_INDEX_TEMPLATE, name, hintMessage), input -> {
+            final Integer selectIndex = validateUserInputAsInteger(input, 1, entities.size(),
+                    String.format(PROMPT_MESSAGE_BAD_INDEX_TEMPLATE, input, name, hintMessage));
+            return entities.get(selectIndex - 1);
+        });
+    }
+
+    public Integer promoteInteger(String name, Integer defaultValue, int minValue, int maxValue, boolean isRequired) throws IOException {
+        final boolean hasDefaultValue = defaultValue != null;
+        if (!isRequired && hasDefaultValue) {
+            throw new IllegalArgumentException("There is no way to input empty value for a non-required field with default value.");
+        }
+        final String defaultValueMessage = hasDefaultValue ? " (" + TextUtils.blue(defaultValue.toString()) + ")" : "";
+        final String hintMessage = String.format("[%d-%d]%s", minValue, maxValue, defaultValueMessage);
+        final String message = String.format("Please input the %s %s: ", name, hintMessage);
+        System.out.print(message);
+        System.out.flush();
+
+        return loopInput(defaultValue, hasDefaultValue, isRequired, message, input -> {
+            return validateUserInputAsInteger(input, minValue, maxValue, String.format(PROMPT_MESSAGE_NUM_TEMPLATE, input, name, hintMessage));
+        });
+    }
+
+    static class UserInputException extends RuntimeException {
+        private static final long serialVersionUID = 7986173597822766589L;
+
+        UserInputException(String message) {
+            super(message);
+        }
+    }
+
+    private <T> T loopInput(T defaultValue, boolean hasDefaultValue, boolean isRequired, String emptyPromoteMessage,
+            Function<String, T> handleInput) throws IOException {
+        while (true) {
+            final String input = reader.readLine();
+            if (StringUtils.isBlank(input)) {
+                if (hasDefaultValue || !isRequired) {
+                    return defaultValue;
+                }
+                System.out.print(emptyPromoteMessage);
+            } else {
+                try {
+                    return handleInput.apply(input);
+                } catch (UserInputException ex) {
```

Comment:
Shall we use Exception to notify input error here? For it may impact the performance, shall we return null for validate error and return only if the result!=null?

Reply:
We need to return a value or an error, null cannot distinguish them, can use a tuple for two information.